zero/svd2.py

```diff
@@ -0,0 +1,77 @@
+import numpy as np
+
+from zero.recommendation_algorithm import (RecommendationAlgorithm,
+                                           register_algorithm)
+from scipy.sparse import csr_matrix, diags
+from scipy.sparse.linalg import svds
+
+
+def remove_mean(sp_matrix):
+    '''
+    Matrix tricks to remove the mean from a sparse matrix.
```

Comment:
Should be more specific: this does not remove the mean, it removes the mean of the non-zero elements per row (I think — I'm not entirely sure how the `np.diff(sp_matrix.indptr)` works).

Also: if we are doing a copy of the `sp_matrix` anyways… I believe the following is equivalent and more readable:

```python
mask = csr_matrix((np.ones_like(sp_matrix.data), sp_matrix.indices, sp_matrix.indptr), shape=sp_matrix.shape)
row_counts = mask.sum(axis=1).A1
row_counts[row_counts == 0] = 1  #  `row_counts += row_counts == 0` if you feel evil.
means = row_sums / row_counts
return sp_matrix - diags(means) * mask, means
```

Reply:
Hi, sorry I don't think the `indptr` is more readable.
So I think the `row_counts = mask.sum(axis=1).A1` is indeed better; so we remove every occurrence of `indices` and `indptr`.